presto-spi/src/main/java/io/prestosql/spi/type/TypeSignature.java

```diff
@@ -160,7 +163,13 @@ public boolean equals(Object o)
     @Override
     public int hashCode()
     {
-        return Objects.hash(base.toLowerCase(Locale.ENGLISH), parameters);
+        if (hashComputed) {
```

Comment:
There's a data race here due to lack of synchronization between these two variables. It's possible for a thread to observe `hashComputed == false` and `hashCode` being set to 0.

The only safe way to do this is:

```
int hash = hashCode;
if (hash == 0) {
    hash = Objects.hash(...);
    hashCode = hash
}

return hash;
```

Note that this relies on the atomicity semantics for unsynchronized access to `int` fields.

Reply:
Also you need to hande `Objects.hash` returning zero
```java
int hash = hashCode;
if (hash == 0) {
    int tempHash = Objects.hash(...);
    if (tempHash == 0) {
        tempHash++;
    }
    hashCode = hash
}

return hash;